lib/pbench/test/unit/server/test_datasets_list.py

```diff
@@ -0,0 +1,156 @@
+            provide_metadata: Dataset metadata fixture
+        """
+
+        def query_api(
+            payload: JSON, username: str, expected_status: HTTPStatus
+        ) -> requests.Response:
+            token = self.token(client, server_config, username)
+            response = client.get(
+                f"{server_config.rest_uri}/datasets/list",
+                headers={"authorization": f"bearer {token}"},
+                query_string=payload,
+            )
+            assert response.status_code == expected_status
+            return response
+
+        return query_api
+
+    def token(self, client, config: PbenchServerConfig, user: str) -> str:
+        response = client.post(
+            f"{config.rest_uri}/login", json={"username": user, "password": "12345"},
+        )
+        assert response.status_code == HTTPStatus.OK
+        data = response.json
+        assert data["auth_token"]
+        return data["auth_token"]
+
+    def get_results(self, name_list: List[str]) -> JSON:
+        """
+        Translate a list of names into a list of expected results of the
+        abbreviated form returned by `datasets/list`: name, controller,
+        run_id, and metadata.
+
+        Args:
+            name_list: List of dataset names
+
+        Returns:
+            JSON list of dataset values
+        """
+        list: List[JSON] = []
+        for name in sorted(name_list):
+            dataset = Dataset.query(name=name)
+            list.append(
+                {
+                    "name": dataset.name,
+                    "controller": dataset.controller,
+                    "run_id": dataset.md5,
+                    "metadata": {
+                        "dataset.created": f"{dataset.created:%Y-%m-%d:%H:%M}"
+                    },
+                }
+            )
+        return list
+
+    @pytest.mark.parametrize(
+        "login,query,results",
+        [
+            ("drb", {"name": "fio"}, ["fio_1", "fio_2"]),
+            ("drb", {"owner": "drb"}, ["drb", "fio_1"]),
+            ("drb", {"controller": "foobar"}, []),
+            ("drb", {"name": "drb"}, ["drb"]),
+            ("test", {"name": "drb"}, []),
+            ("test_admin", {"name": "drb"}, ["drb"]),
+            ("drb", {"controller": "node"}, ["drb"]),
+            ("drb", {}, ["drb", "fio_1", "fio_2"]),
+            ("test", {}, ["test", "fio_1", "fio_2"]),
+            ("test_admin", {}, ["drb", "test", "fio_1", "fio_2"]),
+            ("drb", {"start": "2000-01-01", "end": "2005-12-31"}, ["fio_2"]),
+            ("drb", {"start": "2005-01-01"}, ["drb", "fio_1"]),
+            ("drb", {"end": "2020-09-01"}, ["drb", "fio_1", "fio_2"]),
+            ("drb", {"end": "1970-09-01"}, []),
+        ],
+    )
+    def test_dataset_list(self, query_as, login, query, results):
+        query.update({"metadata": ["dataset.created"]})
+        result = query_as(query, login, HTTPStatus.OK)
+        assert result.json == self.get_results(results)
```

Comment:
The call at line 138 is slightly magical:  it makes the call to `query_as()` return the value that the call to `get_results()` expects...which feels like a layering or encapsulation violation to me...but I can't come up with a better suggestion.  😕

At the very least, a code comment in `get_results()` explaining the requirement would be greatly helpful.  (And, one here as well would be good, too.)

Reply:
I wouldn't call it "magical" ... I was just trying to minimize the bulk of the parametrization by factoring out the common expression. But there's no docstring here, and I should have written one to clarify.